src/main/java/org/elasticsearch/search/aggregations/reducers/movavg/models/HoltWintersModel.java

```diff
@@ -0,0 +1,421 @@
+                for (SeasonalityType policy : values()) {
+                    validNames.add(policy.getName());
+                }
+                throw new ElasticsearchParseException("Invalid seasonality type: [" + text + "], accepted values: " + validNames);
+            }
+            return result;
+        }
+
+        private final byte id;
+        private final ParseField parseField;
+
+        SeasonalityType(byte id, String name) {
+            this.id = id;
+            this.parseField = new ParseField(name);
+        }
+
+        /**
+         * Serialize the SeasonalityType to the output stream
+         *
+         * @param out
+         * @throws IOException
+         */
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeByte(id);
+        }
+
+        /**
+         * Deserialize the SeasonalityType from the input stream
+         *
+         * @param in
+         * @return    SeasonalityType Enum
+         * @throws IOException
+         */
+        public static SeasonalityType readFrom(StreamInput in) throws IOException {
+            byte id = in.readByte();
+            for (SeasonalityType seasonalityType : values()) {
+                if (id == seasonalityType.id) {
+                    return seasonalityType;
+                }
+            }
+            throw new IllegalStateException("Unknown Seasonality Type with id [" + id + "]");
+        }
+
+        /**
+         * Return the english-formatted name of the SeasonalityType
+         *
+         * @return English representation of SeasonalityType
+         */
+        public String getName() {
+            return parseField.getPreferredName();
+        }
+    }
+
+
+    /**
+     * Controls smoothing of data. Alpha = 1 retains no memory of past values
+     * (e.g. random walk), while alpha = 0 retains infinite memory of past values (e.g.
+     * mean of the series).  Useful values are somewhere in between
+     */
+    private double alpha;
+
+    /**
+     * Equivalent to <code>alpha</code>, but controls the smoothing of the trend instead of the data
+     */
+    private double beta;
+
+    private double gamma;
+
+    private int period;
+
+    private SeasonalityType seasonalityType;
+
+    private boolean pad;
+    private double padding;
+
+    public HoltWintersModel(double alpha, double beta, double gamma, int period, SeasonalityType seasonalityType, boolean pad) {
+        this.alpha = alpha;
+        this.beta = beta;
+        this.gamma = gamma;
+        this.period = period;
+        this.seasonalityType = seasonalityType;
+        this.pad = pad;
+
+        // Only pad if we are multiplicative and padding is enabled
+        // The padding amount is not currently user-configurable...i dont see a reason to expose it?
+        this.padding = seasonalityType.equals(SeasonalityType.MULTIPLICATIVE) && pad ? 0.0000000001 : 0;
+    }
+
+
+    @Override
+    public boolean hasNext(int windowLength) {
```

Comment:
Can we name this slightly differently? When `hasNext()` returns false it is usually (in iterators) an indication that your should stop iterating because you have reached the end, but in this case we are saying that 'there isn't a value for this data point but you should keep going because there might be values for subsequent data points'. Maybe `hasValue()` would be better?